Schema evolution based on ROOT and Reflex dictionaries

hegner opened the pull request:
BEGINRELEASENOTES
- Allow comparison of data schemata across versions
- Provide syntax to clarify user intentions in schema evolution
- Provide schema evolution implementation based on ROOT backend
- Include infrastructure for future support for schema evolution in other backends
- Documentation for schema evolution functionality

ENDRELEASENOTES

Schema evolution based on old Reflex dictionary approach.

tmadlener committed:
Remove unused fields

tmadlener committed:
Add SchemaEvolution singleton to hold evolution functions

tmadlener committed:
Inject type information into collection buffers

tmadlener committed:
[wip] Start populating SchemaEvolution registry

tmadlener committed:
Inject current schema version into buffers from buffer factory

tmadlener committed:
[wip] Split registration into two steps

Easier this way to get the current version information into the whole
system

tmadlener committed:
[wip] Require registration of each evolution function

Remove two step registration again

tmadlener committed:
[clang-tidy] Mark inputs as const& for now

tmadlener committed:
Create schema_evolution test subdirectory and build old datamodel

tmadlener committed:
Add first simple tests for "trivial" schema evolution

tmadlener committed:
Fix test environment and typo

tmadlener committed:
Add failing test for renamed member variables

Commit:
Merge branch 'master' into schema-evol-library

Commit:
move Collection::createBuffers template into macro

Commit:
creating components and datatypes for explicit schema evolution

Commit:
add more schema evolution code generation

Commit:
bump of schema version for testing. version 1 is already reserved for non-versioned legacy data

Commit:
add missing schema evolution pieces; prepare for ioread rules in reflex dicts

Commit:
add code generation for reflex schema evolution

hegner commented:
@tmadlener - this is the current status w/ Reflex based schema evolution. It contains a debug output. The problem is that the value of `y_old` read in is already claimed to be zero. Opening with bare ROOT the value however is correct.
If playing with not-evolved members like `x` we get the proper number. Thus my assumption is that there is no proper read before this custom code kicks in. Still - this code crashes if the original file does not contain an `y_old` leaf. So I would assume another ROOT bug behind this.

tmadlener committed:
Reduce unnecessary template instantiations

tmadlener committed:
Fix preprocessor directives

tmadlener committed:
Move function implementations into .cc files for Components

hegner commented:
@tmadlener - seems I never reported back in the tracker. so here for the records. Newer ROOT versions do not suffer from this issue any more.

Commit:
Merge branch 'master' into schema-reflex

Commit:
Update SchemaEvolution.cc

Commit:
Update Collection.cc.jinja2

Commit:
disable currently unused schema evolution parts

Commit:
address static code checker warnings

tmadlener committed:
Fix bug re-introduced in merging master

tmadlener reviewed (commented):
This is working for me locally with ROOT v6.28/06 and as far as the schema evolution part is concerned seems to be complete (TBC at tomorrows meeting). I have left a few inline comments for cleanup / nitpicking.

As we have already discussed in private handling multiple old versions will be done in a separate PR. From a technical point of view that should "just be another loop" over all the older versions.

The roundtrip tests that dump the model from a written file are currently failing because the generation of the dumped model does not generate any schema evolution code, but the diffing against the original model has schema evolution code now:
```console
92: + diff -ru /home/tmadlener/work/AIDASoft/podio/build/tests/root_io/example_frame.root.dumped_datamodel/datamodel /home/tmadlener/work/AIDASoft/podio/tests/datamodel
92: Only in /home/tmadlener/work/AIDASoft/podio/tests/datamodel: ExampleWithArrayv2Data.h
92: Only in /home/tmadlener/work/AIDASoft/podio/tests/datamodel: ExampleWithNamespacev2Data.h
92: Only in /home/tmadlener/work/AIDASoft/podio/tests/datamodel: ExampleWithUserInitv2Data.h
92: Only in /home/tmadlener/work/AIDASoft/podio/tests/datamodel: NamespaceStructv2.h
```

This leads me to a point on the naming and file layout of the generated code, currently we will have a
- `FooData` in `FooData.h`
- `FoovXData` in `FoovXData.h`

I think this could be cleaner if we had a `vX` namespace and potentially a `vX` subfolder to put "old" headers into. (The latter could also make the roundtrip tests slightly easier to fix).

Finally, I think we should consider splitting this PR into two:
- One that only has the refactoring of the jinja templates that happens in the early commits here which leads to quite a bit of unrelated noise here. We would then also avoid having these mixed up with the "interesting" changes when we squash in the end.
- The actual changes for schema evolution.

hegner commented on python/templates/Collection.cc.jinja2:
```diff
@@ -223,6 +193,36 @@ bool registerCollection() {
+{% if old_schema_version is defined %}
+    // register a buffer creation function for the schema evolution buffer
+   // SCHEMA EVOLUTION: Not yet required with only ROOT backend
+   // factory.registerCreationFunc("{{ class.full_type }}Collection", {{ old_schema_version }}, createBuffersV{{old_schema_version}}); //TODO
```
correct.

Commit:
addressing PR comments

hegner commented on python/podio_class_generator.py:
```diff
@@ -264,6 +378,28 @@ def _process_datatype(self, name, definition):
+          iorule.version = self.old_schema_version.lstrip("v")
+          iorule.source = f'{member_type} {schema_change.member_name_old}'
+          iorule.target = schema_change.member_name_new
+          iorule.code = f'std::cout<< "reading {schema_change.member_name_old} with value " << onfile.{schema_change.member_name_old} << std::endl; {iorule.target} = onfile.{schema_change.member_name_old};' # noqa
```
done

hegner commented on python/podio_class_generator.py:
```diff
@@ -248,12 +288,86 @@ def _process_component(self, name, component):
+      self._fill_templates('Component', component)
+      self.root_schema_component_names.add(name + self.old_schema_version)
+
+  def _replaceComponentInPaths(self, oldname, newname, paths):
```
done

hegner commented on python/podio_class_generator.py:
```diff
@@ -248,12 +288,86 @@ def _process_component(self, name, component):
+#          iorule.source = "int y_old"
+#          iorule.code = "std::cout<< onfile.y_old << y << std::endl; y = onfile.y_old;"
+#          if len(self.root_schema_iorules) == 0:
+#            self.root_schema_iorules.add(iorule)
```
correct. removed

hegner commented:
Thanks for checking. I was addressing most of the points now.

Concerning the split into two PRs. The part you want to have in a separate PR is probably commit b65c6eecde5101834497cd71a98881d5fb19bee4

I could cherry pick that one for a new PR if that is what you want.

When it comes to namespaces and directory structure for the old data - most of the work will actually be in CMake, and not in the code generation part.  Do we want to do it in a separate PR?

tmadlener commented:
> Concerning the split into two PRs. The part you want to have in a separate PR is probably commit https://github.com/AIDASoft/podio/commit/b65c6eecde5101834497cd71a98881d5fb19bee4

Yes exactly. Plus this one (to not break things again) https://github.com/AIDASoft/podio/pull/472/commits/6b8fe156308161289e212b67e1bfc2e2d5dea4b4

tmadlener commented on python/podio_class_generator.py:
```diff
@@ -248,12 +288,86 @@ def _process_component(self, name, component):
+      self.root_schema_component_names.add(name + self.old_schema_version)
+
+  def _replaceComponentInPaths(self, oldname, newname, paths):
+    """Replace component in paths"""
```
Ah sorry, I missed the `if` in the loop. I thought all elements were touched, but they aren't. Then we leave it like this.

tmadlener commented:
One other observation: The key4hep-nightlies seem to work for the schema evolution part, but they only come with root 6.28/04, so maybe we don't even need the latest version?

andresailer commented on python/podio_class_generator.py:
```diff
@@ -115,23 +135,24 @@ def __init__(self, yamlfile, install_dir, package_name, io_handlers, verbose, dr
 
-  def process_schema_evolution(self):
+  def preprocess_schema_evolution(self):
     """Process the schema evolution"""
```
```suggestion
    """Preprocess the schema evolution"""
```

andresailer commented on python/podio_class_generator.py:
```diff
@@ -156,6 +178,17 @@ def process_schema_evolution(self):
+      # now go through all the io_handlers and see what we have to do
+      if 'ROOT' in self.io_handlers:
+        for item in root_filter(comparator.schema_changes):
+          schema_evolutions = self.root_schema_dict.get(item.klassname)
```
This should replace the entire loop body
```python
# add whatever is relevant to our ROOT schema evolution
self.root_schema_dict.setdefault(item.klassname, []).append(item)
```
Or use a `defaultdict`

hegner replied:
good observation. simplified now.

andresailer commented on python/podio_class_generator.py:
```diff
@@ -248,12 +288,78 @@ def _process_component(self, name, component):
+        if member.full_type in self.root_schema_dict:
+          needs_schema_evolution = True
+          # prepare the ROOT I/O rule
+          print(member.full_type)
```
remove debugging print?

hegner replied:
yap.

andresailer commented on python/podio_class_generator.py:
```diff
@@ -248,12 +288,78 @@ def _process_component(self, name, component):
+          needs_schema_evolution = True
+          # prepare the ROOT I/O rule
+          print(member.full_type)
+          dir(member)
```
remove?

hegner replied:
yap

tmadlener committed:
Rearrange schema evolution tests to not interfere with others

Make the schema evolution datamodel the new one and use the original one
to write old data. This allows the roundtrip tests to work without
additional work because the additional schema evolution code is not used
anywhere for that.

tmadlener committed:
Add a test for a `float` to `double` migration

hegner commented on python/podio_schema_evolution.py:
```diff
@@ -107,6 +107,17 @@ def __init__(self, name, member_name_old, member_name_new):
     super().__init__(f"'{self.name}': member '{self.member_name_old}' renamed to '{self.member_name_new}'.")
 
 
+class RootIoRule:
```
that actually got moved there after the discussion with Thomas.

Commit:
addressing review comments and code checker

hegner commented:
> One other observation: The key4hep-nightlies seem to work for the schema evolution part, but they only come with root 6.28/04, so maybe we don't even need the latest version?

Any 6.28 is enough. 6.24 is not. And 6.26 I did not try.

tmadlener commented:
Ah alright. Then I simply misunderstood "latest release". 6.26 did not work for me.

hegner commented:
@tmadlener - if you are fine, can you review and approve?

tmadlener committed:
Merge branch 'master' into schema-reflex

tmadlener commented on python/podio_class_generator.py:
```diff
@@ -247,12 +271,76 @@ def _process_component(self, name, component):
+    # Compute and prepare the potential schema evolution parts
+    schema_evolution_datatype = copy.deepcopy(datatype)
+    needs_schema_evolution = False
+    # check whether it has a renamed member
+    # if name in self.root_schema_dict.keys():
+    # for member in schema_evolution_datatype['Members']:
+    # if
+    # then check for components with a renamed member
```
```suggestion
```
Comment seems to no longer correspond with the code that follows.

tmadlener commented on python/podio_class_generator.py:
```diff
@@ -263,6 +351,28 @@ def _process_datatype(self, name, definition):
+          iorule.code = f'{iorule.target} = onfile.{schema_change.member_name_old};'
+          self.root_schema_iorules.add(iorule)
+        else:
+          raise NotImplementedError("Schema evolution for this type not yet implemented")
```
This is potentially confusing, without further information on what "type" is.

tmadlener reviewed (commented):
Looks good. Can you confirm whether the comments I marked are really left overs from earlier? The Error message can be addressed in a follow up PR as well, just wanted to have it on a "list" somewhere.

Commit:
Update python/podio_class_generator.py

Co-authored-by: Thomas Madlener <thomas.madlener@desy.de>
